### ISSUE TITLE
Facility to generate metric of stable inference testing

### ISSUE THREAD

**Issue by dngoldberg** [EdiGlacUQ/fenics_ice#83]
We are currently testing the "leave one [many] out" capability of the inversion to test the result of the inversion on velocity points that are not used in the cost function. I suggest [solver.comp_J_inv](https://github.com/EdiGlacUQ/fenics_ice/blob/0471046fbe102803c23ac668599e045d5d365d59/fenics_ice/solver.py#L1198) be reused/augmented for this purpose. 

Currently comp_J_inv calculates the mismatch cost function on the list of observed velocities and uncertainties, which can be seen as J_mis = \sum_i  (u_i-u_{i,obs})^2/u_std_i^2 + (u_i-u_{i,obs})^2/u_std_i^2 and adds this to the regularisation cost J_reg.

We wish to produce J_mis, but for  a list of "test" velocities, which are distinct from the velocities used for inversion. For simplification, u_std/v_std for this test list should be 1 (enforced either within the model through kwargs or toml parameters, or externally through input file creation). J_reg should be zero. There should not be overlap of velocity points between the set used for inversion and the test set. This can be enforced externally through input file creation, or within model.py (the latter will be more difficult).

We can make use of comp_J_inv to achieve our aim by forcing it to use this "test" list, held in a separate file referenced by the toml. J_reg can be zeroed by setting prior parameters to zero through [this function](https://github.com/EdiGlacUQ/fenics_ice/blob/0471046fbe102803c23ac668599e045d5d365d59/fenics_ice/solver.py#L175).

Through this, comp_J_inv can be called a second time, with the solver parameters set to the results of the inversion -- this can be done with a "chain" of subsampling with a common test data set.

**Comment by bearecinos** [comment]
Ok I am going to expand this comments just for my use later... also I have a few questions for you @dngoldberg  below (shown in **bold**)

**BTW: would this be the right equation? In the one you wrote looks that only u is taken into account.**
![\Large  ](https://latex.codecogs.com/svg.image?&space;J_{mis}&space;=&space;\sum_i&space;&space;\frac{(u_i-u_{i,obs})^2}{u_std_i^2}&space;&plus;&space;\frac{(v_i-v_{i,obs})^2}{v_std_i^2}&space;)

![\Large  ](https://latex.codecogs.com/svg.image?&space;J_{mis}&space;&plus;&space;J_{reg}&space;&space;)

> We wish to produce J_mis, but for a list of "test" velocities, which are distinct from the velocities used for inversion. For simplification, u_std/v_std for this test list should be 1 (enforced either within the model through kwargs or toml parameters, or externally through input file creation).

**This can be done anywhere as we can just divide std value by itself...  u_std/u_std for every point in the array, make a assertions so we make sure all are 1?**

> J_reg should be zero. 

[via this function](https://github.com/EdiGlacUQ/fenics_ice/blob/0471046fbe102803c23ac668599e045d5d365d59/fenics_ice/solver.py#L175).
```
def zero_inv_params(self):
        """Set delta_alpha, gamma_alpha to zero (e.g. for misfit only hessian)"""
        self.delta_alpha = 1E-10
        self.gamma_alpha = 1E-10
        self.delta_beta = 1E-10
        self.gamma_beta = 1E-10

        if self.delta_beta_gnd is not None:
            self.delta_beta_gnd = 1E-10
```
> There should not be overlap of velocity points between the set used for inversion and the test set. This can be enforced externally through input file creation, or within model.py (the latter will be more difficult).

This is already done as we have completely different files of velocity ... one containing only training data and a different file for test data. 

**@dngoldberg  I understand that what we need is a way to make iterative inversions?**.... I suggest we take the approach used by the l-curves; where we have a "toml_template" and generate several tomls ... In this case with different subsample files (for test and training data).

On the new `run_script.py` we then only set up **one testing configuration** for a single pair of toml files (test.toml training.toml) then run that script several times.  **Would this work?**

So to summarize we need

1) Under /runs/ A new script that:

- makes the inversion a set of "training" velocities (from a specific toml) then saves the output of alpha and beta. 
- Takes the output of alpha and beta and calls comp_J_inv(), and force the function to use the "test" list of velocities (from another toml, held in a separate file and sets J_reg can be zero by setting prior parameters to zero either from the toml) .
- Stores the J_mis output.

We can done this recursively for as many subsamples files of training and test data that we generate. 

2) Changes to comp_J_inv() 

- We dont even need to change this! As our velocities will come from the toml!

3) We need things at the Smith_glacier_repo 

- Build a new generation of toml files based on the names of the velocity files: subsample_test and training
- We have two templates:
1."training" > the usual toml template
2. "test" > with the test data but all J_reg terms set to zero ... alpha constant = 0.0, beta from data = 0 and all gamma_alpha, ect =0.

I think this might minimize changes and we actually dont need to change comp_J_inv()

**Comment by dngoldberg** [comment]
> **BTW: would this be the right equation? In the one you wrote looks that only u is taken into account.**
![\Large  ](https://latex.codecogs.com/svg.image?&space;J_{mis}&space;=&space;\sum_i&space;&space;\frac{(u_i-u_{i,obs})^2}{u_std_i^2}&space;&plus;&space;\frac{(v_i-v_{i,obs})^2}{v_std_i^2}&space;)

you are correct. i wrote one component only to save time. Below, i call this sum **J_test**.

> **This can be done anywhere as we can just divide std value by itself...  u_std/u_std for every point in the array, make a assertions so we make sure all are 1?**

I think there are many ways to do this. For instance you could set u_std[:] =1. As long as it does not affect other parts of the workflow.

> **@dngoldberg  I understand that what we need is a way to make iterative inversions?**.... I suggest we take the approach used by the l-curves; where we have a "toml_template" and generate several tomls ... In this case with different subsample files (for test and training data).

yes, i think this is a good way to characterise this. To compare and contrast:

- An l-curve is a set of inversions where the observation data is always the same, but regularisation parameters differ. The important outputs are the cost terms for each inversion of the iteration.
- Our **Stable inference iteration** will be an iteration of inversions where each iteration uses the *same test data set* but a *different observational set* where each observational set is a subsample of the original with a different step value (and not containing any location/value from the test dataset). Outputs will be a differnet J_test for each step.

There was discussion of doing a different L-curve for each step value -- but (a) I think this is not a good use of resources and (b) it confuses the comparison of J_test, so i do not think it should be done.

> So to summarize we need
> -------------------------------
> 
> 1) Under /runs/ A new script that:
> 
> - makes the inversion for a set of "training" velocities (from a specific toml) then saves the output of alpha and beta. 
> - Takes the output of alpha and beta and calls comp_J_inv(), and force the function to use the "test" list of velocities (from another toml, held in a separate file and sets J_reg can be zero by setting prior parameters to zero either from the toml) .
> - Stores the J_mis output.

yes. above i called this J_test to emphasise that it is over the test data set, and not the "training" one.

> 
> We can done this recursively for as many subsamples files of training and test data that we generate. 

yes. i would say iteratively, not recursively.

> 
> 2) Changes to comp_J_inv() 
> 
> - We dont even need to change this! As our velocities will come from the toml!

As of now, comp_J_inv() only knows about the one cloud-point velocity set. At the very least, it needs to be called somehow with the test set. Im not too concerned about how this is done as long as it's done. Personally i feel that the inversion and the evaluation of J_test should be done in the same script -- but this would mean that you would need a single toml for this script, so you cannot, for instance, just change obs.velfile from the toml mid-script, if this is what you are thinking. But we can talk through this later.

> 
> 3) We need things at the Smith_glacier_repo 
> 
> - Build a new generation of toml files based on the names of the velocity files: subsample_test and training
> - We have two templates:
> 1."training" > the usual toml template 2. "test" > with the test data but all J_reg terms set to zero ... alpha constant = 0.0, beta from data = 0 and all gamma_alpha, ect =0.
> 
> I think this might minimize changes and we actually dont need to change comp_J_inv()

see above. I think it will be difficult to have ZERO changes to comp_J_inv() - but they need not be invasive.
